internal/controllers/mic_reconciler.go

```diff
@@ -0,0 +1,336 @@
+		return res, fmt.Errorf("failed tp update the status for MIC %s based on builds: %v", micObj.Name, err)
+	}
+
+	err = r.micReconHelper.processImagesSpecs(ctx, micObj, pods)
+	if err != nil {
+		return res, fmt.Errorf("failed to process images spec: %v", err)
+	}
+	return res, nil
+}
+
+//go:generate mockgen -source=mic_reconciler.go -package=controllers -destination=mock_mic_reconciler.go micReconcilerHelper
+type micReconcilerHelper interface {
+	updateStatusByPullPods(ctx context.Context, micObj *kmmv1beta1.ModuleImagesConfig, pods []v1.Pod) error
+	updateStatusByMBSC(ctx context.Context, micObj *kmmv1beta1.ModuleImagesConfig) error
+	processImagesSpecs(ctx context.Context, micObj *kmmv1beta1.ModuleImagesConfig, pullPods []v1.Pod) error
+}
+
+type micReconcilerHelperImpl struct {
+	client     client.Client
+	podHelper  pullPodManager
+	micHelper  mic.MIC
+	mbscHelper mbsc.MBSC
+	scheme     *runtime.Scheme
+}
+
+func newMICReconcilerHelper(client client.Client,
+	pullPodHelper pullPodManager,
+	micHelper mic.MIC,
+	mbscHelper mbsc.MBSC,
+	scheme *runtime.Scheme) micReconcilerHelper {
+	return &micReconcilerHelperImpl{
+		client:     client,
+		podHelper:  pullPodHelper,
+		mbscHelper: mbscHelper,
+		micHelper:  micHelper,
+		scheme:     scheme,
+	}
+}
+
+func (mrhi *micReconcilerHelperImpl) updateStatusByPullPods(ctx context.Context, micObj *kmmv1beta1.ModuleImagesConfig, pods []v1.Pod) error {
+	logger := ctrl.LoggerFrom(ctx)
+	if len(pods) == 0 {
+		return nil
+	}
+
+	podsToDelete := make([]v1.Pod, 0, len(pods))
+	patchFrom := client.MergeFrom(micObj.DeepCopy())
+
+	for _, p := range pods {
+		image := p.Labels[imageLabelKey]
+		imageSpec := mrhi.micHelper.GetModuleImageSpec(micObj, image)
+		if imageSpec == nil {
+			logger.Info(utils.WarnString("image not present in spec during updateStatusByPods, deleting pod"), "mic", micObj.Name, "image", image)
+			podsToDelete = append(podsToDelete, p)
+			continue
+		}
+		phase := p.Status.Phase
+		switch phase {
+		case v1.PodFailed:
+			if imageSpec.Build != nil || imageSpec.Sign != nil {
+				logger.Info("failed pod with build or sign spec, updating image status to NeedsBulding")
+				mrhi.micHelper.SetImageStatus(micObj, image, kmmv1beta1.ImageNeedsBuilding)
+			} else {
+				logger.Info(utils.WarnString("failed pod without build or sign spec, shoud not have happened"), "image", image)
+			}
+			podsToDelete = append(podsToDelete, p)
+		case v1.PodSucceeded:
+			logger.Info("successful pod, updating image status to ImageExists")
+			mrhi.micHelper.SetImageStatus(micObj, image, kmmv1beta1.ImageExists)
+			podsToDelete = append(podsToDelete, p)
+		}
+	}
+	// patch the status in the MIC object
+	err := mrhi.client.Status().Patch(ctx, micObj, patchFrom)
+	if err != nil {
+		return fmt.Errorf("failed to patch the status of mic %s: %v", micObj.Name, err)
+	}
+
+	errs := make([]error, 0, len(podsToDelete))
+	for _, pod := range podsToDelete {
+		err = mrhi.podHelper.deletePod(ctx, &pod)
```

Comment:
If we are deleting each pod separately in a different client call, then why don't we directly delete each pod instead of appending appending it to the list of pods to be deleted?

Reply:
in case updating the MIC status on line 165 fails ( for any reason), we need those pods to be present in the cluster , in order to be able to execute the next reconciliation loop correctly

Reply:
Makes sense. Can we put your explanation in a comment? I will probably ask the same question each time I see this code :)

Reply:
fixed